File: 2/part2.js
const _ = require('lodash');
const passwords = require('./passwords.json');

const validPasswords = passwords.filter(line => {
  const {first, second, char, password} = parseLine(line);
  const a = password[first - 1] === char;
  const b = password[second - 1] === char;
  const valid = (a && !b) || (!a && b);
  // console.log(`f: ${first}, s: ${second}, c: ${char}, p: ${password}, a: ${password[first - 1]} == ${char}, b: ${password[second - 1]} == ${char}, valid: ${valid}`);
  return valid;
});

console.log(validPasswords.length);

function parseLine(line) {
  const split = line.split(' ');
  const pos = split[0].split('-');
  const first = parseInt(pos[0]);
  const second = parseInt(pos[1]);
  const char = split[1][0];
  const password = split[2];

  return {first, second, char, password};
}

File: 2/hs/part2.hs
import qualified Data.Text as Text
import qualified Data.Text.IO as Text
import Parselib

type PasswordPolicy = (Int, Int, Char, String)

line :: Parser PasswordPolicy
line = do
  first <- natural
  symb "-"
  second <- natural
  space
  c <- letter
  symb ":"
  password <- many letter
  space
  return (first, second, c, password)

parseLine :: String -> PasswordPolicy
parseLine l = case apply line l of
  [(a, "")] -> a
  _ -> error l

valid :: String -> Bool
valid l = do
  let (first, second, char, password) = parseLine l
  let a = char == (password !! (first - 1))
  let b = char == (password !! (second - 1))
  (a && not b) || (not a && b)

main = do
  text <- Text.readFile "passwords.txt"
  let list = map Text.unpack $ Text.lines text
  let fil = filter valid list
  let count = length fil
  print count
